src/main/java/org/anarres/parallelgzip/ParallelGZIPOutputStream.java

```diff
@@ -44,25 +37,35 @@ private static DeflaterOutputStream newDeflaterOutputStream(@Nonnull OutputStrea
         return new DeflaterOutputStream(out, deflater, 512, true);
     }
 
-    private static class Block implements Callable<byte[]> {
+    private static class State {
+        private final Deflater def = newDeflater();
+        private final ByteArrayOutputStreamExposed buf = new ByteArrayOutputStreamExposed(SIZE);
+        private final DeflaterOutputStream str = newDeflaterOutputStream(buf, def);
+    }
+
+    /** This ThreadLocal avoids the recycling of a lot of memory, causing lumpy performance. */
+    private static final ThreadLocal<State> STATE = new ThreadLocal<State>() {
+        @Override
+        protected State initialValue() {
+            return new State();
+        }
+    };
 
-        private static class State {
+    /* Allow write into byte[] directly */
+    private static class ByteArrayOutputStreamExposed extends ByteArrayOutputStream {
+        private ByteArrayOutputStreamExposed(int size) {
+            super(size);
+        }
 
-            private final Deflater def = newDeflater();
-            private final ByteArrayOutputStream buf = new ByteArrayOutputStream(SIZE);
-            private final DeflaterOutputStream str = newDeflaterOutputStream(buf, def);
+        private void writeTo(byte[] buf) {
+            System.arraycopy(this.buf, 0, buf, 0, count);
         }
-        /** This ThreadLocal avoids the recycling of a lot of memory, causing lumpy performance. */
-        private static final ThreadLocal<State> STATE = new ThreadLocal<State>() {
-            @Override
-            protected State initialValue() {
-                return new State();
-            }
-        };
-        public static final int SIZE = 64 * 1024;
+    }
+
+    private static class Block implements Callable<Block> {
         // private final int index;
-        private final byte[] in = new byte[SIZE];
-        private int in_length = 0;
+        private final byte[] buf = new byte[SIZE * 2];
```

Comment:
`N * k` - my intuition tells me that such a k must exist. At least for the gzip algorithm. Proof idea: Each entity is compressed to at most x bytes. Since there are a limited amount of entities in each block such a k must exist.

Let me take a look at you branch here. Do the tests pass for your branch?